### DISCUSSION TITLE
PAL Settings API aIndex argument

### DISCUSSION THREAD

**Discussion by jkcko** [openthread/openthread#9836] [Q&A] [answered]
Please clarify how the aIndex argument is used in the PAL settings.h API.

The comments state that the order of the items associated with a specific key does not need to be maintained when added or deleted, but it is not clear on any further requirements with regards to the the use of the index or storage of the items regarding the index.

If `n` items are added and stored in indexes `0` to `n-1`, and then item with index `i` where `i < n-1` is deleted,
- Can item `i` remain empty such that a subsequent call to `otPlatSettingsGet` on index `i` returns `OT_ERROR_NOT_FOUND`?  
- Or must the item at index `i` be filled in such that `otPlatformSettingsGet` returns `OT_ERROR_NONE` for indexes `0` to `n-2` and `otPlatformSettingsGet` with index `n-1` will return `OT_ERROR_NOT_FOUND`?

**Comment by abtink** [comment, marked as answer]
@jkcko,

> Or must the item at index `i` be filled in such that `otPlatformSettingsGet` returns `OT_ERROR_NONE` for indexes `0` to `n-2` and `otPlatformSettingsGet` with index `n-1` will return `OT_ERROR_NOT_FOUND`?

This is the expected behavior.

The settings key are either single-value or list-based.
- `otPlatSettingsGet()` should allow iterating through all items in the list one by one by incrementing `aIndex`.
- `OT_ERROR_NOT_FOUND` indicates end of the list.
-  The order of items in the list-based key can change after `otPlatSettingsAdd()` or `otPlatSettingsDelete()` calls.
